keras_nlp/models/bloom/bloom_causal_lm_test.py

```diff
@@ -0,0 +1,175 @@
+# limitations under the License.
+
+import contextlib
+from unittest.mock import patch
+
+import keras
+import pytest
+
+from keras_nlp.backend import ops
+from keras_nlp.models.bloom.bloom_backbone import BloomBackbone
+from keras_nlp.models.bloom.bloom_causal_lm import BloomCausalLM
+from keras_nlp.models.bloom.bloom_causal_lm_preprocessor import (
+    BloomCausalLMPreprocessor,
+)
+from keras_nlp.models.bloom.bloom_tokenizer import BloomTokenizer
+from keras_nlp.tests.test_case import TestCase
+
+
+class BloomCausalLMTest(TestCase):
+    def setUp(self):
+        self.vocab = ["<unk>", "<s>", "</s>", "<pad>"]
+        self.vocab += ["!", "air", "Ġair", "plane", "Ġat", "port"]
+        self.vocab = dict([(token, i) for i, token in enumerate(self.vocab)])
+        self.merges = ["Ġ a", "Ġ t", "Ġ i", "Ġ b", "a i", "p l", "n e"]
+        self.merges += ["Ġa t", "p o", "r t", "Ġt h", "ai r", "pl a", "po rt"]
+        self.merges += ["Ġai r", "Ġa i", "pla ne"]
+        self.tokenizer = BloomTokenizer(
+            vocabulary=self.vocab, merges=self.merges
+        )
+        self.preprocessor = BloomCausalLMPreprocessor(
+            self.tokenizer,
+            sequence_length=8,
+        )
+        self.backbone = BloomBackbone(
+            vocabulary_size=self.preprocessor.tokenizer.vocabulary_size(),
+            num_layers=2,
+            num_heads=2,
+            hidden_dim=4,
+            intermediate_dim=16,
+            max_sequence_length=8,
+        )
+        self.init_kwargs = {
+            "preprocessor": self.preprocessor,
+            "backbone": self.backbone,
+        }
+        self.train_data = (
+            [
+                " airplane at airport",
+                " airplane airport",
+            ],
+        )
+        self.input_data = self.preprocessor(*self.train_data)[0]
+
+    def test_causal_lm_basics(self):
+        vocabulary_size = self.tokenizer.vocabulary_size()
+        self.run_task_test(
+            cls=BloomCausalLM,
+            init_kwargs=self.init_kwargs,
+            train_data=self.train_data,
+            expected_output_shape=(2, 8, vocabulary_size),
+        )
+
+    def test_generate(self):
+        causal_lm = BloomCausalLM(**self.init_kwargs)
+        # String input.
+        prompt = "airplane at airport"
+        output = causal_lm.generate(prompt)
+        self.assertTrue(prompt in output)
+        # Int tensor input.
+        prompt_ids = self.preprocessor.generate_preprocess([prompt])
+        causal_lm.preprocessor = None
+        outputs = causal_lm.generate(prompt_ids)
+        # Assert prompt is in output in token id space.
+        self.assertAllEqual(
+            outputs["token_ids"][:, :4],
+            prompt_ids["token_ids"][:, :4],
+        )
+        self.assertAllEqual(
+            outputs["padding_mask"][:, :4],
+            prompt_ids["padding_mask"][:, :4],
+        )
+
+    @pytest.mark.keras_3_only
+    def test_generate_with_bfloat16(self):
```

Comment:
we should now support passing dtype to backbone. let's maybe switch this to the following setup, and no global state mutations.

```
backbone = BloomBackbone.from_config({**self.backbone.get_config(), "dtype": "bfloat16"})
causal_lm = BloomCausalLM(backbone=backbone, preprocessor=self.preprocessor)
```